modules/phase_field/src/kernels/KKSACBulkC.C

```diff
@@ -17,39 +17,117 @@ KKSACBulkC::validParams()
-  params.addRequiredCoupledVar("cb", "b-phase concentration");
+  params.addRequiredCoupledVar("global_cs", "Global concentrations, for example, c, b.");
+  params.addRequiredCoupledVar("eta", "Order parameter");
+  params.addRequiredParam<std::vector<MaterialPropertyName>>(
+      "ci_names",
+      "Phase concentrations. These must have the same order as Fj_names and global_cs, for "
+      "example, c1, c2, b1, b2.");
+  params.addParam<std::vector<MaterialPropertyName>>(
+      "dcidb_names",
+      "The derivative of phase concentration wrt global concentration. They must have the same "
+      "order as Fj_names and ci_names, for example, dc1dc, dc2dc, dc1db, dc2db, db1dc, db2dc, "
+      "db1db, db2db.");
+  params.addParam<std::vector<MaterialPropertyName>>(
+      "dcideta_names",
+      "The derivative of phase concentration wrt order parameter. They must have the same order "
+      "as ci_names, for example, dc1deta, dc2deta, db1deta, db2deta.");
+  params.addRequiredParam<MaterialPropertyName>("L_name", "Allen-Cahn mobility");
   return params;
 }
 
 KKSACBulkC::KKSACBulkC(const InputParameters & parameters)
   : KKSACBulkBase(parameters),
-    _ca_name(getVar("ca", 0)->name()),
-    _ca_var(coupled("ca")),
-    _ca(coupledValue("ca")),
-    _cb_name(getVar("cb", 0)->name()),
-    _cb_var(coupled("cb")),
-    _cb(coupledValue("cb")),
-    _prop_dFadca(getMaterialPropertyDerivative<Real>("fa_name", _ca_name)),
-    _prop_d2Fadca2(getMaterialPropertyDerivative<Real>("fa_name", _ca_name, _ca_name)),
-    _prop_d2Fadcadarg(_n_args)
+    _c_names(coupledComponents("global_cs")),
+    _c_map(getParameterJvarMap("global_cs")),
+    _num_c(coupledComponents("global_cs")),
+    _ci_names(getParam<std::vector<MaterialPropertyName>>("ci_names")),
+    _prop_ci(_num_c),
+    _dcidb_names(getParam<std::vector<MaterialPropertyName>>("dcidb_names")),
+    _prop_dcidb(_num_c),
+    _dcideta_names(getParam<std::vector<MaterialPropertyName>>("dcideta_names")),
+    _prop_dcideta(_num_c),
+    _Fa_name(getParam<MaterialPropertyName>("fa_name")),
+    _first_dFa(_num_c),
+    _d2F1dc1db1(_num_c),
+    _L(getMaterialProperty<Real>("L_name"))
 {
-  // get second partial derivatives wrt ca and other coupled variable
-  for (unsigned int i = 0; i < _n_args; ++i)
-    _prop_d2Fadcadarg[i] = &getMaterialPropertyDerivative<Real>("fa_name", _ca_name, i);
+  // declare _prop_ci to be a matrix for easy reference
+  for (unsigned int m = 0; m < _num_c; ++m)
+  {
+    _prop_ci[m].resize(2);
+
+    for (unsigned int n = 0; n < 2; ++n)
+      _prop_ci[m][n] = &getMaterialPropertyByName<Real>(_ci_names[m * 2 + n]);
+  }
+
+  for (unsigned int m = 0; m < _num_c; ++m)
+  {
+    _prop_dcidb[m].resize(2);
+    _prop_dcideta[m].resize(2);
+
+    for (unsigned int n = 0; n < 2; ++n)
+    {
+      // declare _prop_dcideta. m is the numerator species, n is the phase of the numerator i
+      _prop_dcideta[m][n] = &getMaterialPropertyByName<Real>(_dcideta_names[m * 2 + n]);
```

Comment:
`_prop_dcideta` is the output material property from KKSPhaseConcentrationDerivatives. To the KKSACBulkC kernel, `_prop_dcideta` is just a material property. We don't compute the derivative in this kernel. ci is not an explicit function of eta in the kernel.

Reply:
That doesn't really matter, the point here is to automatically build the derivative property names.

Reply:
I guess I should do the same in the KKSPhaseConcentrationDerivatives class? If the users do not supply the derivative names, how can I pass the derivative values from the material to the kernel? The `getMaterialPropertyDerivative(ci, eta)` will just be zero since the kernel does not know that ci is a function of eta.